pkg/vault/vault_interface.go

```diff
@@ -0,0 +1,321 @@
+	if err != nil {
+		msg := "Error adding credentials to vault to " + path + ":" + err.Error()
+		return errors.New(msg)
+	}
+	return nil
+}
+
+// AddSecretFromStruct constructs a vault secret from the given structure
+func (c *Connection) AddSecretFromStruct(path string, creds interface{}) error {
+	jsonStr, err := json.Marshal(creds)
+	if err != nil {
+		return err
+	}
+	credentialsMap := make(map[string]interface{})
+	if err := json.Unmarshal(jsonStr, &credentialsMap); err != nil {
+		return err
+	}
+	return c.AddSecret(path, credentialsMap)
+}
+
+// Mount mounts a key-value secret provider (kv version 1) to manage the storage of the secrets
+func (c *Connection) Mount(path string) error {
+	body := strings.NewReader(`{"type":"kv-v1"}`)
+	url := c.Address + path
+
+	req, err := http.NewRequest("POST", url, body)
+	if err != nil {
+		msg := "Error creating request to mount vault for user credentials " + url + ":" + err.Error()
+		return errors.New(msg)
+	}
+	req.Header.Set("X-Vault-Token", c.Token)
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		msg := "Error mounting vault for user credentials " + url + ":" + err.Error()
+		return errors.New(msg)
+	}
+	defer resp.Body.Close()
+
+	return nil
+}
+
+// GetIdentitiesVaultAuthPath returns the path where policies for the different identities are written
+// Includes the namespace so that multiple m4d control planes can use the same vault instance.
+func GetIdentitiesVaultAuthPath() string {
+	auth, _ := GetVaultAuthService()
+	//	path := "auth/" + auth + "/" + GetSystemNamespace() + "/identities"
+	path := "auth/" + auth
+	return path
+}
+
+// InitConnection creates a new connection to vault.
+// Note that it assumes that the home path has been mounted during the vault setup.
+func InitConnection(addr string, token string) (CredentialManagerInterface, error) {
+	if os.Getenv("RUN_WITHOUT_VAULT") == "1" {
+		return NewDummyConnection(), nil
+	}
+	conf := &api.Config{
+		Address:    addr,
+		HttpClient: httpClient,
+	}
+
+	client, err := api.NewClient(conf)
+	if err != nil {
+		msg := "Error creating vault client: " + err.Error()
+		return nil, errors.New(msg)
+	}
+
+	// Get the vault token stored in config
+	if token == "" {
+		msg := "No vault token found.  Cannot authenticate with vault."
+		return nil, errors.New(msg)
+	}
+
+	client.SetToken(token)
+
+	return NewConnection(client,addr,token), nil
+}
+
+// GetVaultAuthTTL returns the amount of time the authorization issued by vault is valid
+func GetVaultAuthTTL() string {
+	return os.Getenv("VAULT_TTL")
+}
+
+// GetVaultAuthService returns the authentication service that was chosen for use with vault,
+// and the configuration options for it.
+// Vault support multiple different types of authentication such as java web tokens (jwt), github, aws ...
+func GetVaultAuthService() (string, api.EnableAuthOptions) {
+	auth := os.Getenv("VAULT_AUTH")
```

Comment:
Consider getting auth path as a parameter to this function.

Reply:
done